Source/DafnyLanguageServer/Handlers/DafnyFormatter.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Dafny.LanguageServer.Workspace;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Client.Capabilities;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.Dafny.LanguageServer.Handlers; 
+
+public class DafnyFormatter : DocumentFormattingHandlerBase {
+  private readonly ILogger<DafnyCompletionHandler> logger;
+  private readonly IDocumentDatabase documents;
+
+  public DafnyFormatter(ILogger<DafnyCompletionHandler> logger, IDocumentDatabase documents) {
+    this.logger = logger;
+    this.documents = documents;
+  }
+
+  protected override DocumentFormattingRegistrationOptions CreateRegistrationOptions(DocumentFormattingCapability capability,
+    ClientCapabilities clientCapabilities) {
+    return new DocumentFormattingRegistrationOptions() {
+      DocumentSelector = DocumentSelector.ForLanguage("dafny")
+    };
+  }
+
+  public override async Task<TextEditContainer?> Handle(DocumentFormattingParams request, CancellationToken cancellationToken) {
+    var lastDocument = await documents.GetLastDocumentAsync(request.TextDocument.Uri);
```

Comment:
But I can't use `GetResolvedDocumentAsync` because it does not return the resolved program. It only returns an `IdeState`.

Reply:
Would you mind moving the formatting call to `Compilation.cs` by introducing something like `Task<string> GetFormattedDocument { get; }` there and re-exposing that in `DocumentManager` so it can be used in `DafnyFormatter.cs`?

I'm hoping to make DafnyLanguageServer agnostic to Dafny, and moving `Compilation.cs` to `DafnyCore`, so it can replace some of the logic in DafnyMain, like `BoogiePipelineWithRerun`.